File: bitWise.js
// Definindo as permissões
const PERMISSION_READ = 1;
const PERMISSION_WRITE = 2;
const PERMISSION_DELETE = 4;
const PERMISSION_UPDATE = 8;
const PERMISSION_EXPORT = 16;
const PERMISSION_ADMIN = 32;

//Função que simula buscar no banco de dado as permissoes do usuario
function FindOneUser(user, pass) {
    if (user === "admin" && pass === "admin") return { user: "admin", permission: 63 };
    else if (user === "usuario1" && pass === "senha123") return { user: "usuario1", permission: 3 };
    else if (user === "usuario2" && pass === "senha123") return { user: "usuario2", permission: 15 };
    else return false;
};

// Função para verificar se o usuário tem permissão
function hasPermission(userPermissions, requiredPermission) {
    return (userPermissions & requiredPermission) !== 0;
};

// Função para autenticar o usuário e verificar permissões
function authenticateUser(username, pass) {
    // Neste exemplo, estamos apenas verificando se o usuário existe na base de dados
    const userExists = FindOneUser(username, pass);

    if (userExists) {
        // Simulando recuperação das permissões do usuário do banco de dados
        
        const userPermissions = userExists.permission;

        hasPermission(userPermissions, PERMISSION_ADMIN) ? console.log(`${username} é um administrador.`) : console.log(`${username} não é um administrador.`);
        
        hasPermission(userPermissions, PERMISSION_READ) && console.log(`tem permissão de leitura`);
        hasPermission(userPermissions, PERMISSION_WRITE) && console.log(`tem permissão de escrita`);
        hasPermission(userPermissions, PERMISSION_DELETE) && console.log(`tem permissão de deletar`);
        hasPermission(userPermissions, PERMISSION_UPDATE) && console.log(`tem permissão de atualizar`);
        hasPermission(userPermissions, PERMISSION_EXPORT) && console.log(`tem permissão de exportar`);
    } else {
        console.log(`Usuário ${username} não encontrado.`);
    };
};

// Autenticar e verificar permissões para um usuário
authenticateUser("usuario1", "senha123");
console.log('\n');
authenticateUser("usuario2", "senha123");
console.log('\n');
authenticateUser("admin", "admin");

File: BitWise.java
// Classe para representar informações de usuário
class Usuario {
    private String username;
    private String pass;
    private int permission;

    // Construtor
    public Usuario(String username, String pass, int permission) {
        this.username = username;
        this.pass = pass;
        this.permission = permission;
    }

    // Getters para obter as informações do usuário
    public String getUser() {
        return username;
    }

    public int getPermission() {
        return permission;
    }

    public String getPass() {
        pass = null;
        return pass;
    }
}

// Classe contendo o método findOneUser() para buscar usuário no banco de dados
public class BitWise {
    public static final int PERMISSION_READ = (int) Math.pow(2, 0);
    public static final int PERMISSION_WRITE = (int) Math.pow(2, 1);
    public static final int PERMISSION_DELETE = (int) Math.pow(2, 2);
    public static final int PERMISSION_UPDATE = (int) Math.pow(2, 3);
    public static final int PERMISSION_EXPORT = (int) Math.pow(2, 4);
    public static final int PERMISSION_ADMIN = (int) Math.pow(2, 5);
    
    public static boolean hasPermission(int userPermissions, int requiredPermission){
        return (userPermissions & requiredPermission) != 0;
    }

    public static Usuario findOneUser(String user, String pass) {
        if (user.equals("admin") && pass.equals("admin")) {
            return new Usuario("admin", "admin", PERMISSION_READ+ PERMISSION_WRITE + PERMISSION_DELETE + PERMISSION_UPDATE + PERMISSION_EXPORT + PERMISSION_ADMIN);
        } else if (user.equals("user1") && pass.equals("user1")) {
            return new Usuario("user1", "user1", PERMISSION_READ+ PERMISSION_WRITE);
        }
        return null;
    }
    public static void main(String[] args) {
      
        Usuario user = findOneUser("admin", "admin");

        if (user == null) {
            System.out.println("Usuário não encontrado:");
        } else {
            String result = hasPermission(user.getPermission(), 32) ? user.getUser()+" é um adm" : user.getUser()+" NÃO é um admin";
            System.out.println(result);

            if (hasPermission(user.getPermission(), PERMISSION_READ))
                System.out.println("tem permissão de leitura");
            if (hasPermission(user.getPermission(), PERMISSION_WRITE))
                System.out.println("tem permissão de escrita");
            if (hasPermission(user.getPermission(), PERMISSION_DELETE))
                System.out.println("tem permissão de deleção");
            if (hasPermission(user.getPermission(), PERMISSION_UPDATE))
                System.out.println("tem permissão de atualização");
            if (hasPermission(user.getPermission(), PERMISSION_EXPORT))
                System.out.println("tem permissão de exportar");
        }

    }
}
